listener.go

```diff
@@ -150,23 +150,23 @@ type listener struct {
 	connID   int
 	connLock sync.Mutex
 
-	factory   TLSFactory
-	storage   TLSStorage
-	version   string
-	tlsConfig *tls.Config
-	cert      *tls.Certificate
-	sans      []string
-	maxSANs   int
-	init      sync.Once
+	factory      TLSFactory
+	storage      TLSStorage
+	version      string
+	tlsConfig    *tls.Config
+	cert         *tls.Certificate
+	certIsSetSig chan struct{}
+	sans         []string
+	maxSANs      int
+	init         sync.Once
 }
 
 func (l *listener) WrapExpiration(days int) net.Listener {
 	ctx, cancel := context.WithCancel(context.Background())
 	go func() {
-		// busy-wait for certificate preload to complete
-		for l.cert == nil {
-			runtime.Gosched()
-		}
+
+		// wait for cert to be set, this will unblock when the channel is closed
+		<-l.certIsSetSig
```

Comment:
Can we confirm that the channel is wired up and will be closed on all code paths that reach this wait?

Reply:
In `if !cert.IsValidTLSSecret(secret) in loadCert`,  it won't.

Reply:
> Can we confirm that the channel is wired up and will be closed on all code paths that reach this read?

The channel is created in the `NewListner` function, the only location where a `listener` struct is instantiated.
> In if !cert.IsValidTLSSecret(secret) in loadCert, it won't.

In `if !cert.IsValidTLSSecret(secret)` in `loadCert` it won't be closed because `l.cert` will still be nil. Which I believe is the desired behavior and the same behavior seen with the previous implementation. Since`l.cert` is private it can only be set at the end of `loadCert`.

Reply:
Is it possible that the `cert` may be load multi times ? 

  if true, then the `WrapExpiration` is not very reasonable, it seems should be triggered right after each successful `loadCert`, as you should run `WrapExpiration` from the beginning, it check first then decide long or short timer;  but the current code, the `WrapExpiration` will still be in laster timer

  if not, then it's still reasonable to trigger `WrapExpiration` in `loadCert`, only after a true load happens, you dont need a channel

whenever a `cert` is scucessfully loaded, stop the potential previous `WrapExpiration`, and start a new `WrapExpiration`

Reply:
`l.cert` can be updated after it is initially set by setting the version to empty string `l.version = ""`, as seen [here](https://github.com/rancher/dynamiclistener/blob/master/listener.go#L384-L387). Updating the cert would not require us to start a new `WrapExperation`.
> whenever a cert is scucessfully loaded, stop the potential previous WrapExpiration, and start a new WrapExpiration

There are some reasons I would not want to do this.
1. We can not stop `WrapExperation` unless we close the listener, which we do not want to do.
2. `WrapExperation` is public, so we would have to keep track of all calls to the function to perform proper clean up of all go routines. 
3.  I don't think we want to potentially break other projects by changing the function signature of `WrapExperation` to work around 1 & 2
4. By not starting a new instance of `WrapExperation`  we have the chance of updating cert expiration before 6 hours which is an acceptable outcome.